File: staCMRx/staSTATS.m
function output = staSTATS (data, shrink, warning)
% function output = staSTATS (data, shrink, warning)
% returns summary statistics on data
% data is a ngroup * nvar cell array of matrices:
% each row of data corresponds to one of ngroup groups of participants
% each column of data corresponds to one of nvar dependent variables (DVs)
% each cell of data is a nsub_ij x nwithin_ij matrix of observations
% where, nsub_ij = no. of participants in group i for variable j,
% nwithin_ij = no. of within-subjects conditions in group i for variable j 
% This means that the total number of conditions (ncond) is sum_i=1:N (nwithin_i1)
% Note: The same participants may or may not be measured under each DV.
% This is currently ignored. Each DV is assumed to be independent of all
% other DVs. Also assumed that each DV is measured under all conditions
% (otherwise STA/SDA is meaningless)
% shrink is option for covariance matrix shrinkage (0=none, 1=diag,
% -1=estimated)
% if warning is set then a warning message is printed if nans are detected
%
% staSTATS returns a 1 * nvar cell array means, cov, nsub and weights
% output{ivar}.means = 1 x ncond vector of observed means for DV ivar
% output{ivar}.n = ncond x ncond matrix of number of subjects
% output{ivar}.cov = ncond x ncond observed covariance matrix
% output{ivar}.regcov = regularized covariance matrix
% output{ivar}.shrinkage = shrinkage parameter for regularization
% output{ivar}.weights = 1 x ncond weight matrix for monotonic regression
% output{ivar}.lm = ncond x ncond matrix of loftus-masson within-subject variance
% output{ivar}.shrink = shrinkage factor (between 0 and 1)
% if shrink==1 then diagonalise weight matrix, if shrink==0 then do nothing
%
% *************************************************************************
% last modified 13 October 2016 (to treat nans as missing)
% *************************************************************************

if nargin < 3, warning = 0; end
if nargin < 2, shrink = -1; end %  estimate shrinkage by default

if ~iscell(data)
    y = gen2cell(data); % convert from general format
else
    y = data; % cell array format
end

ngroup = size(y,1); nvar = size(y,2);

% calculate statistics
output = cell(1,nvar); 
for ivar = 1:nvar
    output{ivar}.means = [];
    output{ivar}.cov = [];
    output{ivar}.regcov = [];
    output{ivar}.n = [];
    output{ivar}.weights = [];
    output{ivar}.lm = [];
    output{ivar}.shrinkage = [];
    output{ivar}.nanflag = [];
    output{ivar}.bad = [];
    for igroup = 1:ngroup
        yy = y{igroup, ivar};
%        u = sum(isnan(yy),2); yy=yy(u==0,:); % delete nans
        out.nanflag = sum(sum(isnan(yy))); % check for nans
        out.means = mean(yy,1,'omitnan'); % means excluding nans
        out.means(isnan(out.means)) = 0; % Means can not be NAN (means whole column is NAN & bad will be 2 later)
        a=yy;a(isnan(yy))=0;a(~isnan(yy))=1; out.n=a'*a; % number of observations
        out.cov = cov(yy,1,'partialrows'); out.cov(isnan(out.cov))=0;  % covariance
        eigCov = eig(out.cov);
        if cond(out.cov) < 1e6 && min(eigCov) > 0
            [out.regcov, out.shrinkage] = shrinkDiag(yy, shrink); % shrink if required
            out.bad = 0;
        else
            [out.regcov, out.shrinkage] = shrinkDiag(yy, 1); % diagonalise bad covariance matrix
            out.bad = 1;
        end
        eigRegCov = eig(out.regcov);
        if cond(out.regcov) > 1e6 || min(eigRegCov) <= 0
            out.regcov = trace(out.regcov)*eye(size(out.regcov,1))/size(out.regcov,1); % make sure all diagonal elements are nonzero
            out.bad = 2;
        end
        outNmin1 = max(out.n,eye(size(out.regcov,1)));
        out.weights =outNmin1.*out.regcov^-1; % weights
        out.lm = loftusmasson(yy); % loftusmasson within subjects variance
% add to output
        output{ivar}.means = [output{ivar}.means, out.means];
        output{ivar}.cov = blkdiag(output{ivar}.cov, out.cov);
        output{ivar}.regcov = blkdiag(output{ivar}.regcov, out.regcov);
        output{ivar}.n = blkdiag(output{ivar}.n, out.n);
        output{ivar}.weights = blkdiag(output{ivar}.weights, out.weights);
        output{ivar}.lm = blkdiag(output{ivar}.lm, out.lm);
        output{ivar}.shrinkage = [output{ivar}.shrinkage, out.shrinkage];
        output{ivar}.nanflag = [output{ivar}.nanflag, out.nanflag];
        output{ivar}.bad = [output{ivar}.bad, out.bad];
    end
    if sum(output{ivar}.nanflag) > 0 && warning ~= 0
        fprintf ('Warning: %d NANs detected for variable %d', sum(output{ivar}.nanflag), ivar);
        fprintf ('. These are treated as missing.\n');
    end
    if output{ivar}.bad > 0 & warning ~= 0
        fprintf ('Bad covariance matrix detected for variable %d. Type = %d\n', ivar, output{ivar}.bad);
    end
end

function [sigma,shrinkage]=shrinkDiag(x,shrink)
% function [sigma,shrinkage]=shrinkDiag(x,shrink)
% x (t*n): t iid observations on n random variables
% sigma (n*n): invertible covariance matrix estimator
%
% Shrinks towards diagonal matrix
% if shrink is specified, then this constant is used for shrinkage

% de-mean returns
t=size(x,1);
meanx=mean(x,1,'omitnan');

% compute sample covariance matrix
sample = cov(x,1,'partialrows'); sample(isnan(sample))=0;

% compute prior
prior=diag(diag(sample)); 

if (nargin < 2 || shrink == -1) % compute shrinkage parameters
  
  % what we call p 
  y=(x-meanx(ones(t,1),:)).^2; y(isnan(y))=0;
  phiMat = y'*y/t-sample.^2; 
  phi=sum(sum(phiMat)); 
  
  % what we call r
  rho=sum(diag(phiMat)); 
  
  % what we call c
  gamma=norm(sample-prior,'fro')^2; 

  % compute shrinkage constant
  kappa=(phi-rho)/gamma;
  shrinkage=max(0,min(1,kappa/t));
    
else % use specified constant
  shrinkage=shrink;
end

% compute shrinkage estimator
sigma=shrinkage*prior+(1-shrinkage)*sample;

function r = loftusmasson(yy)
% returns Loftus & Masson within subjects error based on Mean Square residual
y_cond = repmat(mean(yy,1,'omitnan'), size(yy,1), 1); 
y_subj = repmat(mean(yy,2,'omitnan'), 1, size(yy,2)); 
m = mean(reshape(yy,1,numel(yy)),'omitnan'); 
y_mean = repmat(m, size(yy,1), size(yy,2)); 
ya = yy - y_cond - y_subj + y_mean; ya(isnan(ya))=0; 
ss = sum(sum(ya.*ya));
df = (size(yy,1)-1)*(size(yy,2)-1);
ms_resid = ss/df;
r = diag(repmat(ms_resid,size(yy,2),1));

File: STACMR-R/staSTATS.R
staSTATS <- function(data, varnames=NULL, shrink=-1, warning=0) {
  ## Calculates statistics for state trace analysis
  ##
  ## Args:
  ##   data: A list of lists containing nsub x ncond matrices
  ##   varnames: An optional list of names of within-participant conditions
  ##   shrink: 0=no shrinkage of covariance; 1=maximal shrinkage; -1=estimated optimal shrinkage
  ##   if warning is set then a warning message is printed if NAs are detected
  ##
  ## Returns a list with the following items:
  ##   means: Observed means
  ##   n: Number of subjects
  ##   cov: Observed covariance matrix
  ##   recov: Regularized (shrinked) covariance matrix 
  ##   weights: Weight matrix for monotonic regression
  ##   lm: Loftus-Masson within subjects variance
  ##   shrinkage: shrinkage parameter (estimated or returned)
  ##   nanflag: count of missing values (NAs)
  
  #if (missing(shrink)) {shrink = -1}
  #if (missing(warning)) {warning = 0}
  
  y <- data
  
  if (is(y, "data.frame")) {y = gen2list(y, varnames)} # convert from general to list format if req'd
  ngroup = length(y); nvar = length(y[[1]])
  
  output = vector("list", nvar)
  
  if ('means' %in% names(y[[1]])) {output = y # already in stats form
  } else {
    for(ivar in 1:nvar) {
      i.means = numeric()
      i.n = matrix(0,0,0)
      i.cov = matrix(0,0,0)
      i.regcov = matrix(0,0,0)
      i.shrinkage = numeric()
      i.weights = matrix(0,0,0)
      i.lm = matrix(0,0,0)
      i.bad = matrix(0,0,0)
      i.nanflag = matrix(0,0,0)
      
      for (igroup in 1:ngroup) {
        y.i <- as.matrix(y[[igroup]][[ivar]])
 #       y.i <- y.i[complete.cases(y.i), ] ## delete rows with NAs
        g.nanflag <- sum(is.na(y.i))
        g.means <- colMeans(y.i,na.rm=TRUE)
        a=y.i; a[is.na(y.i)]=0; a[!is.na(y.i)]=1; g.n=t(a)%*%a; a=g.n-1; a[a<0]=0; # no. of observations
        g.cov <- cov(y.i,use='pairwise.complete.obs')*a/g.n; g.cov[is.na(g.cov)]=0 # adjusted covariance
        
        eigcov <- eigen(g.cov)
        if ((kappa(g.cov,2) < 1e6) && (min(eigcov$values) > 0))
          {s <- shrinkDiag(y.i, shrink); g.bad <- 0}
       else
          {s <- shrinkDiag(y.i, 1); g.bad <- 1} # diagonalize ill-conditioned matrix
        g.regcov <- s$sigma; g.shrinkage = s$shrinkage
        eigRegcov <- eigen(g.regcov) # check if positive definite
        if ((kappa(g.regcov,2) > 1e6) ||(min(eigRegcov$values) <= 0))
        {g.regcov <- sum(eigRegcov$values)*diag(nrow(g.regcov))/nrow(g.regcov); g.bad <- 2}
        
        Nmin <- g.n; Nmin[which(diag(nrow(g.regcov)) > g.n)] <- 1;
        g.weights <- Nmin*solve(g.regcov)
        if (length(g.regcov)==1) {g.lm=g.regcov} else {g.lm = LoftusMasson (y.i)}
        
        # add to vectors and matrices
        i.means = c(i.means, g.means)
        i.n = adiag(i.n, g.n) # requires "magic" package
        i.cov = adiag(i.cov, g.cov)
        i.regcov = adiag(i.regcov, g.regcov)
        i.shrinkage = c(i.shrinkage, g.shrinkage)
        i.weights = adiag(i.weights, g.weights)
        i.lm = adiag(i.lm, g.lm)
        i.bad = c(i.bad, g.bad)
        i.nanflag = c(i.nanflag, g.nanflag)
      }
      # add to list
      out = list(i.means, i.n, i.cov, i.regcov, i.shrinkage, i.weights, i.lm, i.nanflag, i.bad)
      output[[ivar]] = out
      names(output[[ivar]]) = c("means", "n", "cov", "regcov", "shrinkage", "weights", "lm", "nanflag", "bad")
      
      if (sum(i.nanflag > 0) && (warning != 0))
        {cat('Warning:', sum(i.nanflag), ' detected for variable',ivar)}
      if (sum(i.bad > 0) && (warning != 0))
        {cat('Warning: Bad covariance matrix detected for variable',ivar, '. Type = ', i.bad)}
    }
  }
  return(output)
}

